crates/core/component/funding/src/component/liquidity_tournament/rewards/mod.rs

```diff
@@ -0,0 +1,133 @@
+use penumbra_sdk_community_pool::component::StateWriteExt as _;
+use penumbra_sdk_dex::{component::LqtRead as _, lp::position};
+use penumbra_sdk_distributions::component::StateReadExt as _;
+use penumbra_sdk_keys::address::ADDRESS_LEN_BYTES;
+use penumbra_sdk_keys::Address;
+use penumbra_sdk_sct::component::clock::EpochRead as _;
+use penumbra_sdk_txhash::TransactionId;
+
+use super::bank::Bank as _;
+use super::votes::StateReadExt as _;
+use crate::component::StateReadExt as _;
+
+mod gauge;
+use gauge::{Gauge, Share};
+
+async fn position_shares(
+    state: impl StateRead,
+    epoch: u64,
+    asset: asset::Id,
+    top_n: usize,
+) -> anyhow::Result<Vec<(Share, position::Id)>> {
+    let mut stream = state.positions_by_volume_stream(epoch, asset)?.take(top_n);
+    let mut total = 0u128;
+    // This will end up containing the volume of each lp.
+    let mut tallies = Vec::new();
+    while let Some(x) = stream.next().await {
+        let (_, lp, volume) = x?;
+        let volume = volume.value();
+        total += volume;
+        tallies.push((Share::from(volume), lp));
+    }
+    // Then divide each volume by the total volume, to get a relevant share.
+    for x in &mut tallies {
+        *x = ((x.0 / Share::from(total))?, x.1);
+    }
+    Ok(tallies)
+}
+
+pub async fn distribute_rewards(mut state: impl StateWrite + Sized) -> anyhow::Result<()> {
+    let current_epoch = state.get_current_epoch().await?;
+    let params = state.get_funding_params().await?;
+
+    // Because of borrow checking shenanigans, we first collect all the votes,
+    // in the form of a big blob of bytes for voter addresses, and individual tuples
+    // for the asset, power tallies.
+
+    // 80 KB of memory should be enough for anybody.
+    let mut all_voter_bytes = Vec::<u8>::with_capacity(ADDRESS_LEN_BYTES * 1000);
+    let mut tallies = Vec::new();
+    let mut vote_receipts = state.vote_receipts(current_epoch.index);
+    while let Some(x) = vote_receipts.next().await {
+        let (asset, power, voter_bytes) = x?;
+        all_voter_bytes.extend_from_slice(&voter_bytes);
+        tallies.push((asset, power));
+    }
+    // Now actually tally things up.
+    let mut gauge = Gauge::empty();
+    for ((asset, power), voter) in tallies
+        .into_iter()
+        .zip(all_voter_bytes.chunks_exact(ADDRESS_LEN_BYTES))
+    {
+        gauge.tally(asset, power, voter);
+    }
+    let finalized = gauge.finalize(
+        params.liquidity_tournament.gauge_threshold,
+        usize::try_from(params.liquidity_tournament.max_delegators)?,
+    );
+
+    // Get the initial budget, and immediately withdraw it from the community pool.
+    let initial_budget = state
+        .get_lqt_reward_issuance_for_epoch(current_epoch.index)
+        .await
+        .unwrap_or_default();
+    state
+        .community_pool_withdraw(Value {
+            asset_id: *STAKING_TOKEN_ASSET_ID,
+            amount: initial_budget,
+        })
+        .await?;
```

Comment:
If we're not able to debit the community pool, we can just short-circuit any minting and not crash the node. This will be helpful on testnet and devnets, and avoids a kerfuffles with governance param updates.

Reply:
wrt to this, my thinking was just that we take the whole function call in the end of epoch handler, and just log an error if it fails for any reason, continuing execution.